src/main/java/io/bf2/kafka/authorizer/CustomAclAuthorizer.java

```diff
@@ -246,6 +247,66 @@ void addAllowedListeners(Map<String, ?> configs) {
+            return new AclBindingFilter(ResourcePatternFilter.ANY, principalFilter);
+        }).collect(Collectors.toList());
+        if (!toDelete.isEmpty()) {
+            log.info("Orphan deletion: deleting orphaned bindings if present, using filter: {}", toDelete);
+            List<CompletableFuture<AclDeleteResult>> deletionResults = delegate.deleteAcls(null, toDelete)
+                    .stream().map(CustomAclAuthorizer::getAclDeleteResultCompletableFuture).collect(Collectors.toList());
+            CompletableFuture<Void> future = CompletableFuture.allOf(deletionResults.toArray(CompletableFuture[]::new));
+            future.whenComplete((unused, throwable) -> {
+                if (throwable == null) {
+                    log.info("Orphan deletion: deleting orphaned bindings completed successfully");
+                } else {
+                    log.error("Orphan deletion: deleting orphaned bindings completed exceptionally", throwable);
+                }
+            });
+        } else {
+            log.warn("Orphan deletion: no managed principals - skipping orphan binding deletion");
+        }
+    }
+
+    private static CompletableFuture<AclDeleteResult> getAclDeleteResultCompletableFuture(CompletionStage<AclDeleteResult> completionStage) {
+        return completionStage.toCompletableFuture().whenComplete((aclDeleteResult, ex) -> {
+            if (ex != null) {
+                log.error("Orphan deletion: An ACL deletion future completed exceptionally", ex);
+            } else if (aclDeleteResult != null) {
+                logAclDeleteResult(aclDeleteResult);
+            } else {
+                log.warn("Orphan deletion: ACL delete result future completed with null exception and result");
+            }
+        });
+    }
+
+    private static void logAclDeleteResult(AclDeleteResult aclDeleteResult) {
+        Optional<ApiException> filterException = aclDeleteResult.exception();
+        if (filterException.isPresent()) {
+            log.error("Orphan deletion: An ACL delete result failed to match ACL filter to delete ACLs", filterException.get());
+        } else if (aclDeleteResult.aclBindingDeleteResults() != null) {
+            for (AclDeleteResult.AclBindingDeleteResult deleteResult : aclDeleteResult.aclBindingDeleteResults()) {
+                AclBinding binding = deleteResult.aclBinding();
+                Optional<ApiException> exception = deleteResult.exception();
+                if (exception.isPresent()) {
+                    log.error("Orphan deletion: Deletion failed for binding {}", binding, exception.get());
+                } else {
+                    log.info("Orphan deletion: Deletion succeeded for binding {}", binding);
+                }
+            }
+        } else {
+            log.warn("Orphan deletion: ACL delete result had no exception and null results");
```

Comment:
+1 to the NPE guard.   just noting: I checked the Kafka code, and I see that `aclBindingDeleteResults` will be a collection on all paths, so we expect never to end up here.